daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala

```diff
@@ -193,14 +214,15 @@ private[lf] object PartialTransaction {
   */
 private[lf] case class PartialTransaction(
     packageToTransactionVersion: Ref.PackageId => TxVersion,
+    contractKeyUniqueness: ContractKeyUniquenessMode,
     submissionTime: Time.Timestamp,
     nextNodeIdx: Int,
     nodes: HashMap[NodeId, PartialTransaction.Node],
     actionNodeSeeds: BackStack[(NodeId, crypto.Hash)],
     consumedBy: Map[Value.ContractId, NodeId],
     context: PartialTransaction.Context,
     aborted: Option[Tx.TransactionError],
-    keys: Map[GlobalKey, Option[Value.ContractId]],
+    keys: Map[GlobalKey, Option[(Value.ContractId, PartialTransaction.VisibleByKey)]],
```

Comment:
What about the alternative of storing the stakeholders together with the contract-id, so that on lookup the authorization rules can be executed?

I believe the current option might be conflating two separate concepts:
1. What is the slice of the ledger state that is accessible to the submitter for command interpretation, i.e., all contracts visbile to `actAs \union readAs`. I believe this has to only be checked outside the engine, as its purpose is to ensure tha the submitter does not learn about data from co-hosted parties. One can see that when thinking about disclosure, where we plan to explicitly expand the amount of state available to interpretation.
2. The authorization rules of fetch and lookupByKey, whose purpose it is to avoid privacy leaks through the ability of a submitter to ask for validation results on lookups touching contracts on which the submitter never got any kind of delegation.

Reply:
Slight clarification on point 1: given the diffrence in fetch vs https://docs.daml.com/daml/reference/contract-keys.html#visiblebykey, we actually define two slices of ledger state:
1. The slice for resolving key lookups: all contracts where one of `actAs \union readAs` is a stakeholder
2. The slice for resolving fetches: the one from 1. extended with all divulged contracts.

I still believe that it would be valuable if we could organize our code such that these checks on ledger state access are the responsibility of the caller of the engine, while the engine ensures that all produced transactions are internally consistent.

Reply:
> What about the alternative of storing the stakeholders together with the contract-id, so that on lookup the authorization rules can be executed?

This is not about authorization. We already validate this correctly for everything (afaik). This is about the visibility check for contract key lookups (at least one reader must be a stakeholder). Note that there is a bug here already, see https://github.com/digital-asset/daml/issues/9454. Working on a fix for that next which will address the TODOS around the Transient visibility.

Let me walk you through what `ptx.keys` tracked before and what it tracks now:
Before: `ptx.keys` tracked successful global key lookups and creates. For key lookups/fetches, the ledger already validates visibility. For creates it doesn’t (since it is not involved) and the engine cannot do it atm (it doesn’t know readAs parties) leading to the bug mentioned above.
After: `ptx.keys` now tracks the keys of all contracts used in the current transaction regardless of whether they come from a fetch by key, a regular fetch, local, …. This means we can easily detect duplicate key errors: If we ever get two different entries in `ptx.keys`, we have an error. It does however mean that the caching `ptx.keys` was used for needs to take visibility into account, if we know about the contract from a regular fetch, we might still need the ledger to check for visibility.

I’d argue that “keys of all contracts used in the current transaction” is a pretty clear concept and is not conflating anything. The visibility is indeed separate and we could track this is a separate map. But I’d rather have the two together, that makes it easier to make sure we don’t forget to look at visibility in places where we have to and avoids them getting out of sync.

Reply:
Thank you for the detailed explanation @cocreature . It makes sense to me now, and I retract my comment.

If you could work your explanation wrt what is being tracked 'keys of all contracts used in the current transaction' into the code somehow, that would likely be helpful for future readers like me :blush:

Reply:
expanded the docs in d610e8818a4f465a0c8cd70900e85070f566bccb